onnx/onnx.in.proto

```diff
@@ -535,10 +539,13 @@ message TensorProto {
     FLOAT8E5M2 = 19;      // follows IEEE 754, supports nan, inf, mostly used for gradients
     FLOAT8E5M2FNUZ = 20;  // follows IEEE 754, supports nan, not inf, mostly used for gradients, no negative zero
 
-    // 4-bit data-types
+    // 4-bit integer data types
     UINT4 = 21;  // Unsigned integer in range [0, 15]
     INT4 = 22;   // Signed integer in range [-8, 7], using two's-complement representation
 
+    // 4-bit floating point data types
+    FLOAT4E2M1 = 23;
+
     // Future extensions go here.
```

Comment:
@xadupre @onnx/sig-archinfra-approvers Moving previous discussion about CICD failures here for better organization.

I looked through all the places that have float8e4m3fn and uint4 to make sure float4e2m1 is added everywhere, but these tests are still failing. :(
```
FAILED onnx/test/test_backend_test.py::OnnxBackendNodeModelTest::test_cast_FLOAT16_to_FLOAT4E2M1_cpu - onnx.onnx_cpp2py_export.shape_inference.InferenceError: [ShapeInferenceError] (op_type:Cast): output has unsupported type tensor(float4e2m1)
FAILED onnx/test/test_backend_test.py::OnnxBackendNodeModelTest::test_cast_FLOAT4E2M1_to_FLOAT16_cpu - onnx.onnx_cpp2py_export.shape_inference.InferenceError: [ShapeInferenceError] (op_type:Cast): input typestr: T1, has unsupported type: tensor(float4e2m1)
FAILED onnx/test/test_backend_test.py::OnnxBackendNodeModelTest::test_cast_FLOAT4E2M1_to_FLOAT_cpu - onnx.onnx_cpp2py_export.shape_inference.InferenceError: [ShapeInferenceError] (op_type:Cast): input typestr: T1, has unsupported type: tensor(float4e2m1)
FAILED onnx/test/test_backend_test.py::OnnxBackendNodeModelTest::test_cast_FLOAT_to_FLOAT4E2M1_cpu - onnx.onnx_cpp2py_export.shape_inference.InferenceError: [ShapeInferenceError] (op_type:Cast): output has unsupported type tensor(float4e2m1)
FAILED onnx/test/test_backend_test.py::OnnxBackendNodeModelTest::test_dequantizelinear_float4e2m1_cpu - onnx.onnx_cpp2py_export.shape_inference.InferenceError: [ShapeInferenceError] (op_type:DequantizeLinear): x typestr: T1, has unsupported type: tensor(float4e2m1)
FAILED onnx/test/test_backend_test.py::OnnxBackendNodeModelTest::test_quantizelinear_float4e2m1_cpu - onnx.onnx_cpp2py_export.shape_inference.InferenceError: [ShapeInferenceError] (op_type:QuantizeLinear): y_zero_point typestr: T2, has unsupported type: tensor(float4e2m1)
```
Is there a way to run the test with an executable so that one can debug into it with gdb?

Reply:
I find this strange too... Is the opset version set right?

Reply:
@justinchuby Anywhere I should specifically set the opset for these tests?

I also noticed an exclusion list in `test_backend_onnxruntime.py`
```
    # The following tests are not supported by onnxruntime.
    backend_test.exclude(
```
that excludes int4/uint4/fp8 tests. Is this related or do I need to add fp4 to the list?

Reply:
Yes. I would add them to the list

Reply:
Still failing after adding to this exclusion list in `test_backend_onnxruntime.py`.
What's the relationship between `test_backend_onnxruntime.py` and `test_backend_test.py`? The latter only defines a dummy backend, but it's also unclear from the names of the failing tests which backend they are using.

Reply:
@xadupre

Reply:
It seems like the test is getting the list of supported dtypes from the previous op schemas that are already moved to `old.cc`. If I add `float4e2m1` as a supported type to those old definitions then the tests can pass.
Any thoughts how these tests choose the opset? Do we need to explicitly specifiy the opset in the test definition (which wasn't necessary previously)?

Reply:
Finally found the reason. I missed some updates to `operator_set.h` so it was not treating opset 23 as the latest opset.
I wonder if it's possible to add a warning when some op schemas in `defs.cc` is not compiled, in order to prevent errors like this.

Reply:
There are still some errors in CI/CD:

- On windows platform, I'm seeing gateway access violation errors. Something wrong with the windows machines?
- On Ubuntu and MacOS, there are a lot of schema erros like this coming from ops that I didn't touch: 
```
Schema error: Trying to register schema with name Abs (domain:  version: 1) from file /Users/runner/work/onnx/onnx/onnx/defs/math/old.cc line 2743, but it is already registered from file /Users/runner/work/onnx/onnx/onnx/defs/math/old.cc line 2743
```

Also, I wonder if I should wait for the 1.17 branchout to happen first and for the new opset to be open for business on the `main` branch. cc @roborags who is the 1.17 release manager.

Reply:
The windows build issues have disappeared.
However, the issue with the debug build still persists which complains about multiple inclusions. As @justinchuby pointed out it was previously encountered here: https://github.com/onnx/onnx/pull/6079#issuecomment-2073524058

@onnx/sig-archinfra-approvers @onnx/sig-operators-approvers Any suggestions on the root cause or workaround?